### PR TITLE
[Enhancement] [Reporting] [ANZ] Allow section payload to be empty [LOYAL-1784]

### PR DESCRIPTION
**Pull request by andrebalazs55**
https://kaligo.atlassian.net/browse/LOYAL-1784

Background:
The PR is bout to handle the case when we set a section as optional and sending an empty array.
The original Slither design allowed us not to send a key in the payload when the section was optional, but did not allow to send a key with empty value.

Question:
Does any of our applications which using slither depending on this behaviour?

### CODE REVIEW

File: lib/slither/generator.rb [hunk 1]
```diff
@@ -8,20 +8,22 @@ def generate(data)
...
+        if content.any?
           content.each do |row|
             @builder << section.format(row)
           end
-        else
-          unless section.optional
-            raise(Slither::RequiredSectionEmptyError, "Required section '#{section.name}' was empty.")
-          end
         end
       end
```
Comment (marfoldi) [inline lib/slither/generator.rb]: Wdyt of

```ruby
@definition.sections.each do |section|
  content = data[section.name]
  raise(Slither::RequiredSectionEmptyError, "Required section '#{section.name}' was empty.") if !content&.any? && !section.optional

  content = [content] if content.is_a?(Hash)
  content.each do |row|
    @builder << section.format(row)
  end
end
```

**Review by marfoldi — CHANGES_REQUESTED** [review]
Think it's a bit hard to follow what's our intention was in the `generator`, suggested a change, please check if it makes sense.

File: lib/slither/generator.rb (hunk 1 above)
Comment (andrebalazs55) [reply lib/slither/generator.rb]: The iteration on the content should be safe navigated, because we do not want to raise when content is empty, but it will raise when want to call each on nil.
Also I would use none? instead of !any?

Comment (marfoldi) [reply lib/slither/generator.rb]: > The iteration on the content should be safe navigated, because we do not want to raise when content is empty, but it will raise when want to call each on nil.

Niiice 👍 

> Also I would use none? instead of !any?

Think we can't use that with the safe navigation, cuz if we're passing `nil`, then it will be `nil && true` or `nil && false` and because it's a falsey value, we won't raise.

Comment (marfoldi) [reply lib/slither/generator.rb]: Added in https://github.com/Kaligo/slither/pull/3/commits/775d452592a05724cd969bf5a21586931985d40f and bumped to `master`.

**Review by marfoldi — APPROVED** [review]

File: spec/generator_spec.rb [hunk 2]
```diff
@@ -31,14 +33,29 @@
...
+        lambda { @generator.generate(@data) }.should raise_error(Slither::RequiredSectionEmptyError, "Required section 'body' was empty.")
+      end
     end
 
-    it "should raise an error if there is no data for a required section" do
-      @data[:body] = []
-      lambda {  @generator.generate(@data) }.should raise_error(Slither::RequiredSectionEmptyError, "Required section 'body' was empty.")
+    context "when body and header are optional" do
+      let(:optional) { true }
+      it "should raise an error if there is no data for a required section" do
```
Comment (gohkhoonhiang) [inline spec/generator_spec.rb]:
```suggestion
      it "should not raise an error if there is no data for a required section" do
```

File: spec/generator_spec.rb [hunk 3]
```diff
@@ -31,14 +33,29 @@
...
+        @data.delete :header
+        lambda { @generator.generate(@data) }.should_not raise_error
+      end
+
+      it "should raise an error if the data is empty for a required section" do
```
Comment (gohkhoonhiang) [inline spec/generator_spec.rb]:
```suggestion
      it "should not raise an error if the data is empty for a required section" do
```

**Review by gohkhoonhiang — APPROVED** [review]
Spotted some typos. Otherwise LGTM 👍

**Comment by andrebalazs55** [comment]
Screenshot 2021-07-14 at 16 23 14
Screenshot 2021-07-14 at 16 21 55
Screenshot 2021-07-14 at 16 21 29

**Comment by andrebalazs55** [comment]
I could not come up with better idea to test it 🤣
